src/game.py

```diff
@@ -0,0 +1,207 @@
+import pygame
+from pygame.time import Clock
+from threading import Semaphore
+from typing import Optional
+from threading import Thread
+
+from src.client.game_client import Client
+from src.map.game_map import Map
+from src.players.player_factory import PlayerFactory
+from src.players.player import Player
+
+
+class Game(Thread):
+    def __init__(self, name: str = None, max_players: int = 1, num_turns: int = None, is_full: bool = False) -> None:
+        super().__init__()
+
+        self.__name: str = name
+        self.map: Optional[Map] = None
+        self.running: bool = True
+        self.__winner: Optional[int] = None
+        self.__is_full: bool = is_full
+
+        self.num_turns: int = num_turns
+        self.current_turn: Optional[int] = None
+        self.num_rounds: Optional[int] = None
+        self.current_round: Optional[int] = None
+        self.__round_started: bool = False
+
+        self.__current_client: Client = Client()
+        self.__all_clients: dict[Player, Client] = {}
+
+        self.__current_player: Optional[Player] = None
+        self.__waiting_players: list[Player] = []
+        self.__players_in_game: dict[int, Player] = {}
+        self.__player_wins: dict[int, int] = {}
+
+        self.__current_player_idx: Optional[int] = None
+        self.__max_players: int = max_players
+        self.__game_players: int = 0
+        self.__all_players: int = 0
+
+        self.__turn_played_sem: Semaphore = Semaphore(0)
+        self.__clock: Clock = Clock()
+
+    def add_player(self, name: str, password: str = None, is_observer: bool = None) -> None:
+        if self.__game_players >= self.__max_players:
+            is_observer = True
+
+        if not is_observer:
+            self.__game_players += 1
+
+        self.__all_players += 1
+
+        player: Player
```

Comment:
it is not customary to do in python, it is better to specify the type annotation with the initialization